File: redhawk-codegen/redhawk/codegen/jinja/python/ports/templates/generic.uses.py
#{#
# This file is protected by Copyright. Please refer to the COPYRIGHT file
# distributed with this source distribution.
#
# This file is part of REDHAWK core.
#
# REDHAWK core is free software: you can redistribute it and/or modify it under
# the terms of the GNU Lesser General Public License as published by the Free
# Software Foundation, either version 3 of the License, or (at your option) any
# later version.
#
# REDHAWK core is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
# details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program.  If not, see http://www.gnu.org/licenses/.
#}
#% set className = portgen.className()
class ${className}(${component.baseclass.name}.${portgen.templateClass()}):
    def __init__(self, parent, name):
        self.parent = parent
        self.name = name
        self.outConnections = {}
        self.port_lock = threading.Lock()

    def getConnectionIds(self):
        return self.outConnections.keys()

    def _evaluateRequestBasedOnConnections(self, __connection_id__, returnValue, inOut, out):
        if not __connection_id__ and len(self.outConnections) > 1:
            if (out or inOut or returnValue):
                raise PortCallError("Returned parameters require either a single connection or a populated __connection_id__ to disambiguate the call.", self.getConnectionIds())

        if len(self.outConnections) == 0:
            if (out or inOut or returnValue):
                raise PortCallError("No connections available.", self.getConnectionIds())
            else:
                if __connection_id__:
                    raise PortCallError("The requested connection id ("+__connection_id__+") does not exist.", self.getConnectionIds())
        if __connection_id__ and len(self.outConnections) > 0:
            foundConnection = False
            for connId, port in self.outConnections.items():
                if __connection_id__ == connId:
                    foundConnection = True
                    break
            if not foundConnection:
                raise PortCallError("The requested connection id ("+__connection_id__+") does not exist.", self.getConnectionIds())

    def connectPort(self, connection, connectionId):
        self.port_lock.acquire()
        try:
            port = connection._narrow(${portgen.corbaClass()})
            self.outConnections[str(connectionId)] = port
        finally:
            self.port_lock.release()

    def disconnectPort(self, connectionId):
        self.port_lock.acquire()
        try:
            self.outConnections.pop(str(connectionId), None)
        finally:
            self.port_lock.release()

    def _get_connections(self):
        self.port_lock.acquire()
        try:
            return [ExtendedCF.UsesConnection(name, port) for name, port in self.outConnections.iteritems()]
        finally:
            self.port_lock.release()
#{% for operation in portgen.operations() %}

#{% set arglist = ['self'] + operation.args %}
#{%  if arglist %}
#{%   if operation.is_attribute %}
    def ${operation.base_attribute}(self, __connection_id__ = ""):
#{%    if operation.returns %}
        return self.${operation.name}(__connection_id__)
#{%    else %}
        self.${operation.name}(__connection_id__)
#{%    endif %}

#{%   endif %}
    def ${operation.name}(${arglist|join(', ')}, __connection_id__ = ""):
#{%  else %}
#{%   if operation.is_attribute %}
    def ${operation.base_attribute}(__connection_id__ = ""):
#{%    if operation.returns %}
        return self.${operation.name}(__connection_id__)
#{%    else %}
        self.${operation.name}(__connection_id__)
#{%    endif %}

#{%   endif %}
    def ${operation.name}(__connection_id__ = ""):
#{%  endif %}
#{% if operation.returns|length > 1 %}
        retVal = []
#{% elif operation.returns|first == 'string' %}
        retVal = ""
#{% elif operation.returns %}
        retVal = None
#{% endif %}
        self.port_lock.acquire()

#{% set hasreturn = (operation.hasreturnType != 'void') %}
#{% if hasreturn %}
#{%     set returnstate='True' %}
#{% else %}
#{%     set returnstate='False' %}
#{% endif %}
#{% set hasout = operation.hasout %}
#{% if hasout %}
#{%     set _hasout='True' %}
#{% else %}
#{%     set _hasout='False' %}
#{% endif %}
#{% set hasinout = operation.hasinout %}
#{% if hasinout %}
#{%     set _hasinout='True' %}
#{% else %}
#{%     set _hasinout='False' %}
#{% endif %}

        try:
            self._evaluateRequestBasedOnConnections(__connection_id__, ${returnstate}, ${_hasinout}, ${_hasout})
            for connId, port in self.outConnections.items():
                if (__connection_id__ and __connection_id__ != connId):
                    continue
                if port != None:
                    try:
                        ${"retVal = " if operation.returns}port.${operation.name}(${operation.args|join(', ')})
                    except Exception:
                        self.parent._baseLog.exception("The call to ${operation.name} failed on port %s connection %s instance %s", self.name, connId, port)
                        raise
        finally:
            self.port_lock.release()
#{% if operation.returns %}

        return retVal
#{% endif %}
#{% endfor %}

File: redhawk-codegen/redhawk/codegen/jinja/cpp/ports/templates/generic.uses.cpp
/*#
 * This file is protected by Copyright. Please refer to the COPYRIGHT file 
 * distributed with this source distribution.
 * 
 * This file is part of REDHAWK core.
 * 
 * REDHAWK core is free software: you can redistribute it and/or modify it 
 * under the terms of the GNU Lesser General Public License as published by the 
 * Free Software Foundation, either version 3 of the License, or (at your 
 * option) any later version.
 * 
 * REDHAWK core is distributed in the hope that it will be useful, but WITHOUT 
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or 
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License 
 * for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License 
 * along with this program.  If not, see http://www.gnu.org/licenses/.
 #*/
//% set classname = portgen.className()
//% set vartype = portgen.interfaceClass() + '_var'
PREPARE_ALT_LOGGING(${classname},${component.userclass.name})
${classname}::${classname}(std::string port_name, ${component.baseclass.name} *_parent) :
Port_Uses_base_impl(port_name)
{
    parent = static_cast<${component.userclass.name} *> (_parent);
    recConnectionsRefresh = false;
    recConnections.length(0);
}

${classname}::~${classname}()
{
}
/*{% for operation in portgen.operations() %}*/
//% set hasreturn = operation.returns != 'void'
/*{% if hasreturn %}*/
/*{%     set returnstate='true' %}*/
/*{% else %}*/
/*{%     set returnstate='false' %}*/
/*{% endif %}*/
//% set hasout = operation.hasout
/*{% if hasout %}*/
/*{%     set _hasout='true' %}*/
/*{% else %}*/
/*{%     set _hasout='false' %}*/
/*{% endif %}*/
//% set hasinout = operation.hasinout
/*{% if hasinout %}*/
/*{%     set _hasinout='true' %}*/
/*{% else %}*/
/*{%     set _hasinout='false' %}*/
/*{% endif %}*/
/*{%  if operation.readwrite_attr %}*/
${operation.returns} ${classname}::${operation.name}() {
    return _get_${operation.name}("");
}

${operation.returns} ${classname}::_get_${operation.name}(const std::string __connection_id__)
/*{%  else %}*/
/*{%   if operation.arglist %}*/
${operation.returns} ${classname}::${operation.name}(${operation.arglist}, const std::string __connection_id__)
/*{%   else %}*/
${operation.returns} ${classname}::${operation.name}(const std::string __connection_id__)
/*{%   endif %}*/
/*{%  endif %}*/
{
/*{% if hasreturn %}*/
    ${operation.temporary} retval${' = %s' % operation.initializer if operation.initializer};
/*{% endif %}*/
    std::vector < std::pair < ${vartype}, std::string > >::iterator i;

    boost::mutex::scoped_lock lock(updatingPortsLock);   // don't want to process while command information is coming in

    __evaluateRequestBasedOnConnections(__connection_id__, ${returnstate}, ${_hasinout}, ${_hasout});
    if (this->active) {
        for (i = this->outConnections.begin(); i != this->outConnections.end(); ++i) {
            if (not __connection_id__.empty() and __connection_id__ != (*i).second)
                continue;
            try {
                ${"retval = " if hasreturn}((*i).first)->${operation.name}(${operation.argnames});
            } catch (...) {
                LOG_ERROR(${classname},"Call to ${operation.name} by ${classname} failed");
                throw;
            }
        }
    }

/*{% if hasreturn %}*/
/*{%   if operation.temporary.endswith('_var') %}*/
    return retval._retn();
/*{%   else %}*/
    return retval;
/*{%   endif %}*/
/*{% endif %}*/
}
/*{% endfor %}*/

std::string ${classname}::getRepid() const
{
    return ${portgen.interfaceClass()}::_PD_repoId;
}
